packages/graph/hash_graph/lib/graph/src/api/web/mod.rs

```diff
@@ -0,0 +1,99 @@
+//! The Axum webserver for accessing the Graph API datastore operations.
+//!
+//! Handler methods are grouped by routes that make up the `web_api`. See [`data_type`] for a
+//! grouped router example
+
+mod api_resource;
+mod data_type;
+mod error;
+mod property_type;
+
+use axum::{routing::get, Extension, Json, Router};
+use utoipa::{openapi, Modify, OpenApi};
+
+use self::api_resource::RoutedResource;
+use crate::datastore::Datastore;
+
+fn api_resources<T: Datastore>() -> Vec<Router> {
+    vec![
+        data_type::DataTypeResource::routes::<T>(),
+        property_type::PropertyTypeResource::routes::<T>(),
+    ]
+}
+
+fn api_documentation() -> Vec<openapi::OpenApi> {
+    vec![
+        data_type::DataTypeResource::documentation(),
+        property_type::PropertyTypeResource::documentation(),
+    ]
+}
+
+pub fn web_api_with_datastore<T: Datastore>(datastore: T) -> Router {
+    // All api reosuces are merged together into a super-router.
+    let merged_routes = api_resources::<T>()
+        .into_iter()
+        .fold(Router::new(), axum::Router::merge);
+
+    // OpenAPI documentation is also generated by merging resources
+    let open_api_doc = OpenApiDodumentation::openapi();
+
+    // super-router can then be used as any other router.
+    merged_routes
+        // Make sure extensions are added at the end so they are made available to merged routers.
+        .layer(Extension(datastore))
+        .route(
+            "/api-doc/openapi.json",
+            get({
+                let doc = open_api_doc;
+                move || async { Json(doc) }
+            }),
+        )
+}
+
+#[derive(OpenApi)]
+#[openapi(
+        tags(
+            (name = "Graph", description = "HASH Graph API")
+        ),
+        modifiers(&MergeAddon)
+    )]
+struct OpenApiDodumentation;
+
+struct MergeAddon;
+
+impl Modify for MergeAddon {
+    fn modify(&self, openapi: &mut utoipa::openapi::OpenApi) {
+        let api_documentation: Vec<_> = api_documentation();
+
+        let api_components = api_documentation
+            .iter()
+            .cloned()
+            .filter_map(|api_docs| {
+                api_docs
+                    .components
+                    .map(|components| components.schemas.into_iter())
+            })
+            .flatten();
+
+        let mut components = openapi.components.take().unwrap_or_default();
+        components.schemas.extend(api_components);
+        openapi.components = Some(components);
+
+        let mut tags = openapi.tags.take().unwrap_or_default();
+        tags.extend(
+            api_documentation
+                .iter()
+                .cloned()
+                .filter_map(|api_docs| api_docs.tags)
+                .flatten(),
```

Comment:
```suggestion
                .flat_map(|api_docs| api_docs.tags),
```
(Untested)

Reply:
For both this and the other case of `filter_map` + `flatten` the return types are both `Option`s, so we need to seperate step 

(unless we can do 2-level flattening with `flat_map`. Perhaps we can turn the inner Option<Vec<_>> into a Vec<_>, although I don't think we need to collect just for the sake of that)